design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+```
+And for the Direct Grant the flow would be for example:
+```
+Auth type                         | Options
+---------------------------------------------------------------------------
+Username Validation
+Authentication factor (1)
+  |- Password
+Authentication factor (2)          [x] enabled  [ ] disabled  [ ] optional
+  |- OTP                           [x] enabled  [ ] disabled
+  |- Other                         [x] enabled  [ ] disabled
+```
+
+##### Configurable flows
+
+Configurable flows work as is currently the case in Keycloak with the possibility
+to add new authentication flows and either replace the default flow in _Bindings_
+or in the _Client > [client] > Authentication Flow Overrides_. Here the normal configuration of
+an authentication factor is to add (or remove) authenticator executions. However, within an
+authentication execution's configuration, it is possible to set a **disable** flag.
+This is for when an admin doesn't want to keep an execution active, but doesn't want it's
+configuration to be deleted. An **optional** flag is present in the configuration of the
+authentication factor, meaning that it will only be required for users which have at least
+one credential set for an execution that is within that authentication factor.
+
+For UX purposes, the GUI shows directly in the flow when an authentication factor is set
+to optional, or if an authentication execution is disabled.
+
+#### Bindings
+
+No changes are necessary here.
+
+#### Required Actions
+
+Current required actions are a bit too limited to support multi-tokens. While it is currently
+possible to require a user to, for example, register an OTP device, it must be possible for the
+administrator to require "register a valid 1st factor" and "register a valid 2nd factor".
+
+To do this, it must be possible to register a required action multiple times, and to configure
+required actions. It must also be possible to remove required actions from the list. The code
+of the required action should define if it can be registered multiple times, and how it can
+be configured. However, all required actions should be able to be removed, and removing a
+required action should also remove it from all users for which it is currently required.
+
+The purpose of this is to be able to have a "register credential" required action in which
+authentication categories can be set. For example, if an admin were to configure a 2nd factor
+with OTP and Fido2, he could then set a "register credential" with the OTP and Fido2
+authentication categories. A user, when confronted with that required action, would only have
+to register one of the two.
+
+
+#### Policies
+
+The password policy for the realm should remain, but the OTP policy should disappear, as a user
+could have several alternative OTP devices with different settings. As such, the information
+about the OTP settings should probably move to information associated with the credential.
```

Comment:
Not necessarily. You could have an TOTP token with 6 digits, and another TOTP token with 8 digits. In this case the configuration is linked to the token itself. However, for other cases maybe a policy could be global for an entire authentication category, but I can't think of a specific use case for one.

Reply:
I see. There are two kinds of settings for authentication method.
Some settings are to be associated with credentials and it is better to let it user configurable, and some are to be associated with authentication categories and leave it admins only.
As for TOTP, number of digits is the former as you said, and I think look ahead of window is the latter. Even if let the look ahead of window user configurable, authentication category wide policy for look ahead of window should be remained.

Reply:
Isn't the OTP policy a set of security constraints imposed by administrators? In general, I think users to follow the policies defined by the organization in regards to authentication.

Reply:
The configuration of the OTP policy for the realm should be configured in the OTP authenticator, not the realm as it is today. Further, the policy would need to be associated with the users credentials. Otherwise, if the realm policy changes current user credentials may no longer work.

I don't think there is a need for users themselves to be able to change the settings. If that is needed that would make the authenticator more complex. The admin would have to be able to define the acceptable options and the user would have to select from the acceptable options when registering. If this is needed it may be better to register multiple OTP authenticators with different policies allowing users to select from different authenticators.

Reply:
And there is a use-case for allowing users to register multiple OTP devices. A single user may for example have a personal phone and a work phone (or a phone and a tablet).

Reply:
> The configuration of the OTP policy for the realm should be configured in the OTP authenticator, not the realm as it is today. Further, the policy would need to be associated with the users credentials. Otherwise, if the realm policy changes current user credentials may no longer work.

I agree. This is an issue we have today that we need to fix, so we need to have OTP policy associated with the user credentials. This will help administrators to change their policies without impact users and opens the door for enhancements such as setting required actions to the users so that they can be told to update their OTP accordingly.

I think I see the reasoning behind setting OTP policy in the authenticator, but how we are going to handle situations where there are different flows with distinct OTP policies/authenticator and the same user authenticating through these flows? The user would have to know in advance which OTP to use depending on the flow he is going through. Isn't this a UX issue?

> I don't think there is a need for users themselves to be able to change the settings. If that is needed that would make the authenticator more complex. The admin would have to be able to define the acceptable options and the user would have to select from the acceptable options when registering. If this is needed it may be better to register multiple OTP authenticators with different policies allowing users to select from different authenticators.

I agree that users should not be able to change settings but rely on whatever policy that was defined by the administrator.

I don't think users are in a position to choose the OTP policy they want to use, but use whatever policy the administrator has chosen as the most secure for their deployments. If an administrator defines OTP with both 6 and 8 digits, why let users choose the weaker?

Reply:
> And there is a use-case for allowing users to register multiple OTP devices. A single user may for example have a personal phone and a work phone (or a phone and a tablet).

By "phone" I guess you are talking about software tokens. If so, I still don't see a reason to "register" multiple OTP devices.

On the other hand, hardware OTP tokens are really being replaced by either software or security keys as 2nd factor, such as Yubico.

Also, is it safe to allow the user to use his personal phone and software token where the OTP policy is weaker than the one enforced by his company in his work phone?

Reply:
With two factor authentication you need at least one backup option. Personally I have a phone and a tablet, where I use the tablet as backup (both devices are just as secure as each other, well the tablet is more secure as it rarely leaves the house ;)). Of course alternatives could be SMS, email, backup codes, etc.. However, why should we limit it? It's up to the admin of the realm and the user to choose what they want, not us.

Reply:
Because, IMO, this does not represent a real use case. How many people using Keycloak today is demanding multiple OTP registrations (I'm talking about software OTP tokens)?

Credential-wise, an OTP is not different than a password but something that you have. 

I agree with you that we are missing some recovery method for 2nd factors though.

Reply:
> I think I see the reasoning behind setting OTP policy in the authenticator, but how we are going to handle situations where there are different flows with distinct OTP policies/authenticator and the same user authenticating through these flows? The user would have to know in advance which OTP to use depending on the flow he is going through. Isn't this a UX issue?

The idea of the multi-token is that the user can select which device to use. Since he can label the device, he can choose "work phone" or "home tablet" (taking Stian's example), the user doesn't need to guess which device he should take, as any which keycloak is allowing at this point is valid. I don't know for how many people this is a use case or not, but I don't see why not allow it however, as it seems convenient without introducing any security risk. And if we are anyway introducing the possibility for a user to have multiple alternative 1st or 2nd factors, the mechanisms to handle it will already be there. 

For policy, I'm not to sure what to make of the discussion up until now. If all flows for a realm of type OTP must be the same, then it makes sense to attach the policy to some thing global, like the realm as is currently the case, or the authentication category (and if so, I'll make a proposition on how this could work at the same time as I describe more explicitly what an authentication category would be). However, if you want to be able to have multiple OTP strengths within the realm, depending on the flow, or even attached to authentication level as @mposolda has suggested in another comment, then it must be attached to the configuration of the authentication execution. 
In either case, since the credential records its characteristics (number of characters, hash algorithm used, otp type, token period), keycloak shouldn't allow the selection of an OTP device that doesn't match the current policy during login.

Reply:
That is true for hardware tokens but I don't think it applies to software tokens. I'm not saying we should not have multiple devices per user, we need that indeed. But that we only need a single OTP software token per user. Just like it is today.

For policy, yes, I agree. I'm just not convinced that users should be able to choose between OTPs if they are different in terms of policy. Seems odd to me ...

Reply:
I think that long term, if we want to have true flexibility with regards to credentials and types, we may need a possibility for people to introduce their own "credential types" as well as "credential policy types" . Rather than support just few "concrete" policies for specific credential types "password" and "totp" . In other words, maybe something like this:

- Under left tab "Authentication" in admin console have subtab "Credentials" .

- Admin can create instances of Credentials based on available Credential Provider Types. Available Credential providers will be by default: password, otp, WebAuthn

- By default, there will be 1 instance of "password", 1 instance of OTP and 1 instance of WebAuthn created.

- In case that credential provider supports credential provider policies (which both password and OTP obviously support), there will be another UI page shown when admin can configure credential policies for given credential instance

- In other words, admin can for example have setup with 1 instance of "password" credential and 2 instances of "otp" credential called "otp-credential-1" and "otp-credential-2" where each of the OTPs will be configured with different OTP policies

- In the authentication flow configuration, I can "link" the specified authenticator execution (or authentication category?) with the given credential instance. In other words, in the "browser" flow, I can add "OTP Form" authenticator and point to "otp-credential-1" . In the DirectGrant flow, I will create "ValidateOTP" authenticator and point it to "otp-credential-1" as well. This will allow me to share same TOTP policies for both the "browser" and "direct grant" flow as policies are configured for the instance of "otp-credential-1" 

- Migration from previous version should be fine as after migration, we will have 1 instance of "password" credential with password policies, which are in current version configured at realm level. Same for TOTP .

Reply:
This seems a good and flexible approach to credentials in Keycloak. This probably deserves its own design document.